### DISCUSSION TITLE
A custom `transport` passed to `@supabase/realtime-js > RealtimeClient` cannot not access internal `_onConnOpen`, `_onConnError`, `_onConnMessage`, `_onConnClose`.

### DISCUSSION THREAD

**Discussion by liyachun01** [supabase/supabase#30361] [Questions]
Using realtime-js, If a custom `transport` is passed to `@supabase/realtime-js > RealtimeClient`, it cannot access internal `_onConnOpen`, `_onConnError`, `_onConnMessage`, `_onConnClose`. While the connection is successful, messages won't be correctly resolved.

Looking at the source code, the `this.setupConnection()` is missing. Is it intentional (Let the custom transport resolve message) or is it a bug?

https://github.com/supabase/realtime-js/blob/82e996685da8ecb75f327d7faf52096b5953c3c3/src/RealtimeClient.ts#L192-L202
```typescript
 // supabase/realtime-js/src/RealtimeClient.ts#L192-L202
if (this.transport) {
  this.conn = new this.transport(this._endPointURL(), undefined, {
    headers: this.headers,
  })
  return
}
if (NATIVE_WEBSOCKET_AVAILABLE) {
  this.conn = new WebSocket(this._endPointURL())
  this.setupConnection()
  return
}
```

https://github.com/supabase/realtime-js/blob/82e996685da8ecb7Ï5f327d7faf52096b5953c3c3/src/RealtimeClient.ts#L418-L427
```typescript
 // supabase/realtime-js/src/RealtimeClient.ts#L418-L427
private setupConnection(): void {
  if (this.conn) {
    this.conn.binaryType = 'arraybuffer'
    this.conn.onopen = () => this._onConnOpen()
    this.conn.onerror = (error: WebSocketLikeError) =>
      this._onConnError(error as WebSocketLikeError)
    this.conn.onmessage = (event: any) => this._onConnMessage(event)
    this.conn.onclose = (event: any) => this._onConnClose(event)
  }
}
```

**Comment by lauri865** [comment]
Facing the same issue. Seems like an oversight to me - how can you possibly implement a custom transport like this.
